replace macvlan with veth pairs

Omarabdul3ziz opened the pull request:
### Description
replace macvlan with veth pairs

### Related Issues
- https://github.com/threefoldtech/zos/issues/2403

List of related issues

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstring

Omarabdul3ziz committed:
replace macvlan with veth pairs in zdb workloads

Co-authored-by: Ashraf Fouda <ashraf.m.fouda@gmail.com>

iwanbk commented on pkg/network/networker.go:
```diff
@@ -156,28 +157,54 @@ func (n *networker) attachYgg(id string, netNs ns.NetNS) (net.IPNet, error) {
+			return netlink.RouteAdd(&route)
+		}); err != nil {
+			return net.IPNet{}, fmt.Errorf("failed to setup link addresses %s: %w", ZDBYggIface, err)
+		}
```
Looks like this code block does the same thing as https://github.com/threefoldtech/zos/pull/2414/files#diff-4e15b6f2760eb52c8292eadf92ce4a6ba3afcf08d34b528d56081b9bc1623554R252-R285

extract it to a func?

Omarabdul3ziz committed:
save progress: working zdb through ygg/my

Co-authored-by: Ashraf Fouda <ashraf.m.fouda@gmail.com>

Omarabdul3ziz committed:
Merge branch 'main' of https://github.com/threefoldtech/zos into main_replace_macvlan

Omarabdul3ziz committed:
chore: arrange and separate the code

ashraffouda commented on pkg/network/ifaceutil/interface.go:
```diff
@@ -179,6 +217,40 @@ func MakeVethPair(name, master string, mtu int) (netlink.Link, error) {
 }
 
+// moveLinkToNamespace moves the veth link to the specified namespace and renames it
+func moveLinkToNamespace(veth netlink.Link, tmpName, finalName string, netNs ns.NetNS, peerLink netlink.Link) (netlink.Link, error) {
```
i suggest we make tmpName > name the underlying method shouldn't know about tmpName I think
and finalName >  renameTo which would be a pointer because it maybe nil if we need to just move it to the name space

ashraffouda commented on pkg/network/networker.go:
```diff
@@ -210,62 +203,120 @@ func (n *networker) attachMycelium(id string, netNs ns.NetNS) (net.IPNet, error)
+		return fmt.Errorf("failed to set link address: %w", err)
+	}
+
+	// if err := options.Set(name, options.IPv6Disable(false)); err != nil {
```
remove unneeded commented coder

ashraffouda commented on pkg/network/networker.go:
```diff
@@ -210,62 +203,120 @@ func (n *networker) attachMycelium(id string, netNs ns.NetNS) (net.IPNet, error)
+	return netlink.LinkSetUp(link)
+}
+
+func createInterface(ifName, bridge string, netNs ns.NetNS, ip *net.IPNet, route *netlink.Route) error {
```
rename it to CreateVethInterface

Omarabdul3ziz committed:
- unify setting the link up step
- create the tmp name inside the veth creation
- rename the function/argument

muhamadazmy commented on pkg/network/ifaceutil/interface.go:
```diff
@@ -133,25 +134,60 @@ func RandomName(prefix string) (string, error) {
+		return nil, fmt.Errorf("master link: %s not found: %v", master, err)
+	}
+
+	nsName := filepath.Base(netNs.Path())
```
What if netNs is `nil` I see you already calling the function with `nil` as a last argument already in some places

muhamadazmy commented on pkg/network/ifaceutil/interface.go:
```diff
@@ -133,25 +134,60 @@ func RandomName(prefix string) (string, error) {
+	ifName := name
+	if netNs != nil {
+		ifName = tmpName
+	}
```
Would be easier to read if it's re-written like
```go
ifName := name
if netNs != nil {
   ifName, err = ip.RandomVethName()
   if err != nil {
       return nil, err
   }
}
```

muhamadazmy commented on pkg/network/ifaceutil/interface.go:
```diff
@@ -161,24 +197,67 @@ func MakeVethPair(name, master string, mtu int) (netlink.Link, error) {
+	// either from the host, or from the namespace
+	var veth2 netlink.Link
+
+	veth2, err = netlink.LinkByName(name)
```
Does a link with name `name` even exist here? Note that so far the 2 links that exists are `ifName` and `peer`, not `name`!

Also `veth2` is not a good varible name. i don't know what you refer too.

muhamadazmy commented on pkg/network/ifaceutil/interface.go:
```diff
@@ -161,24 +197,67 @@ func MakeVethPair(name, master string, mtu int) (netlink.Link, error) {
+	if err := netlink.LinkSetUp(peerLink); err != nil {
+		return nil, fmt.Errorf("failed to set veth peer up: %w", err)
+	}
+
```
I don't believe this function should return anything because the `veth` can be inside a namespace. So returning a link that is inside a namespace does not make any sense. Normally to work with links that lives inside a namespace you have to do 
```go
ns.Run(func() {
  // load the nic here
  // work with the nic
}
```

muhamadazmy commented on pkg/network/networker.go:
```diff
@@ -156,28 +157,21 @@ func (n *networker) attachYgg(id string, netNs ns.NetNS) (net.IPNet, error) {
-			},
-			Gw: gw.IP,
-			// LinkIndex:... this is set by macvlan.Install
+	route := netlink.Route{
```
I think even if this `now` has only a single route. It's important to understand that this won't always be the case. It's totally legal to configure multiple routes and taking away this ability is not recommended.

muhamadazmy reviewed (commented):
thanks @Omarabdul3ziz for the good work :) I left few comments below that might be helpful

Omarabdul3ziz committed:
- allow multiple routing addresses for ygg/myc
- refetch the created link when needed instead of returning it from the creation method
- organize and enhance naming

Omarabdul3ziz committed:
- simplify the peer naming
- validate the peer absence

Omarabdul3ziz committed:
comment unused createMacVlan func
